crates/solver/src/settlement_rater.rs

```diff
@@ -244,18 +248,251 @@ impl SettlementRating for SettlementRater {
             None => Score::Protocol(big_rational_to_u256(&objective_value).unwrap_or_default()),
         };
 
+        // recalculate score if success probability is provided
+        let score = match settlement.success_probability {
+            Some(success_probability) => {
+                match self.score_calculator.compute_score(
+                    &objective_value,
+                    &inputs.gas_cost,
+                    success_probability,
+                ) {
+                    Ok(score) => score,
+                    Err(err) => {
+                        tracing::warn!(?err, "Failed to compute score with success probability");
+                        score
+                    }
+                }
+            }
+            None => score,
+        };
+
+        let gas_cost = match settlement.gas_cost.as_ref() {
+            Some(gas_cost) => GasCost::SolverEstimated(gas_cost.to_big_rational()),
+            None => {
+                let gas_cost = &simulation.gas_estimate.to_big_rational() * &effective_gas_price;
+                GasCost::ProtocolEstimated(gas_cost)
+            }
+        };
+
         let rated_settlement = RatedSettlement {
             id,
             settlement,
             surplus: inputs.surplus_given,
             earned_fees,
             solver_fees: inputs.solver_fees,
             gas_estimate: simulation.gas_estimate,
-            gas_price: effective_gas_price.clone(),
+            gas_price: effective_gas_price,
+            gas_cost,
             objective_value,
             score,
             ranking: Default::default(),
         };
         Ok(Rating::Ok(rated_settlement))
     }
 }
+
+pub struct ScoreCalculator {
+    score_cap: BigRational,
+    submission_strategies: Vec<TransactionStrategyArg>,
+}
+
+impl ScoreCalculator {
+    pub fn new(score_cap: BigRational, submission_strategies: Vec<TransactionStrategyArg>) -> Self {
+        Self {
+            score_cap,
+            submission_strategies,
+        }
+    }
+
+    pub fn cost_fail(&self, gas_cost: &BigRational) -> BigRational {
+        if self.submission_strategies == vec![TransactionStrategyArg::Flashbots] {
+            BigRational::zero()
+        } else {
+            gas_cost.clone()
```

Comment:
Can we add a comment here why we assume the full cost? I was thinking we should use 1/2 as usually we don't fully execute the settlement?

Reply:
Something like: "The cost in case of a revert can deviate non-deterministically from the cost in case of success and it is often significantly smaller."?
If we include a scaling of 1/2 we should add more info on how that number came to be.